Changes in module 'ebi-metagenomics/krona/ktimporttext'
--- modules/ebi-metagenomics/krona/ktimporttext/main.nf
+++ modules/ebi-metagenomics/krona/ktimporttext/main.nf
@@ -4,7 +4,7 @@
 
 process KRONA_KTIMPORTTEXT {
     tag "$meta.id"
-    label 'process_single'
+    label 'very_light'
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
@@ -24,6 +24,11 @@
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+    
+    if (meta.containsKey("var_region")){
+        prefix = prefix + "_${meta.var_region}"
+    }
+
     """
     ktImportText  \\
         $args \\

--- modules/ebi-metagenomics/krona/ktimporttext/meta.yml
+++ modules/ebi-metagenomics/krona/ktimporttext/meta.yml
@@ -17,7 +17,10 @@
       documentation: http://manpages.ubuntu.com/manpages/impish/man1/ktImportTaxonomy.1.html
       tool_dev_url: https://github.com/marbl/Krona
       doi: 10.1186/1471-2105-12-385
-      licence: ["https://raw.githubusercontent.com/marbl/Krona/master/KronaTools/LICENSE.txt"]
+      licence:
+        [
+          "https://raw.githubusercontent.com/marbl/Krona/master/KronaTools/LICENSE.txt",
+        ]
 input:
   - meta:
       type: map

************************************************************
